File: 2021/src/day12/day12.py
class Cave:
    def __init__(self, name):
        self.start = name == "start"
        self.end = name == "end"
        self.name = name
        self.small = name.lower() == name
        self.connections = set()

    def paths_to_end(self, inc_visits, visited_small_twice=False):
        visits = inc_visits.copy()
        visits[self.name] += 1
        if self.end:
            return [self.name]

        if self.small and visits[self.name] > 1:
            visited_small_twice = True

        check = [cave for cave in self.connections if not cave.start]

        if visited_small_twice:
            check = [
                cave
                for cave in check
                if (cave.small and visits[cave.name] == 0) or (not cave.small)
            ]

        if len(check) > 0:
            paths = []
            for cave in check:
                path = [self.name]
                path += cave.paths_to_end(visits, visited_small_twice)
                paths.append(path)

            return paths
        else:
            return [self.name, ""]


def get_caves():
    pairs = open("input.txt", "r").read().splitlines()
    caves = {}
    for pair in pairs:
        [a, b] = pair.split("-")
        if not a in caves:
            caves[a] = Cave(a)
        if not b in caves:
            caves[b] = Cave(b)
        caves[a].connections.add(caves[b])
        caves[b].connections.add(caves[a])
    return caves


def path_collapse(paths, path_list, str=""):
    for path in paths:
        # print(path)
        if type(path) is list:
            path_collapse(path, path_list, str)
        else:
            if path == "end":
                str += path
                path_list.append(str)
            else:
                str += path + " -> "
    return path_list


def part1():
    caves = get_caves()
    visits = {}
    for cave in caves:
        visits[cave] = 0

    paths = caves["start"].paths_to_end(visits, True)
    paths = path_collapse(paths, [])

    print("Part 1")
    print("   ", len(paths))


def part2():
    caves = get_caves()
    visits = {}
    for cave in caves:
        visits[cave] = 0

    paths = caves["start"].paths_to_end(visits, False)
    paths = path_collapse(paths, [])

    print("Part 2")
    print("   ", len(paths))


part1()
part2()

File: 2021/src/day12/day12.js
class DefaultMap {
  constructor(defaultValue) {
    this._default = defaultValue;
    this._map = new Map();
  }

  get(key) {
    this._map.get(key) ?? this._default;
  }

  has(key) {
    return this._map.has(key);
  }

  set(key, value) {
    return this.set(key, value);
  }
}

class Cave {
  constructor(name) {
    this.name = name;
    this.end = name === "end";
    this.small = name.toLowerCase() === name;
    this.start = name === "start";
    this.connections = new Set();
    Cave._caves.set(name, this);
  }

  static _caves = new Map();

  static emptyVisitMap() {
    const map = new Map();
    for (const [key] of Cave._caves) {
      map.set(key, 0);
    }
    return map;
  }

  static get(name) {
    return Cave._caves.get(name);
  }

  *pathsToEnd({
    oneSmallRevisitAllowed = false,
    visitCount = Cave.emptyVisitMap(),
  } = {}) {
    const visits = new Map(visitCount);
    visits.set(this.name, visits.get(this.name) + 1);

    if (this.end) {
      yield this.name;
      return;
    }

    if (this.small && visits.get(this.name) > 1) {
      oneSmallRevisitAllowed = false;
    }

    let checks = Array.from(this.connections)
      .map((name) => Cave.get(name))
      .filter((cave) => !cave.start);

    if (!oneSmallRevisitAllowed) {
      checks = checks.filter(
        (cave) => (cave.small && visits.get(cave.name) == 0) || !cave.small,
      );
    }

    if (checks.length) {
      for (const check of checks) {
        for (const path of check.pathsToEnd({
          visitCount: visits,
          oneSmallRevisitAllowed,
        })) {
          yield `${this.name} --> ${path}`;
        }
      }
    } else {
      yield `${this.name} --> STOP`;
    }
  }
}

const parseInput = (raw) => {
  const caves = new Map();
  const lines = raw.split("\n");
  for (const line of lines) {
    const [a, b] = line.split("-");
    if (!caves.has(a)) {
      caves.set(a, new Cave(a));
    }
    if (!caves.has(b)) {
      caves.set(b, new Cave(b));
    }
    caves.get(a).connections.add(b);
    caves.get(b).connections.add(a);
  }
  return Cave._caves;
};

export const part1 = (raw) => {
  parseInput(raw);
  const paths = Array.from(Cave.get("start").pathsToEnd()).filter((path) =>
    path.endsWith(" --> end"),
  );
  return paths.length;
};

export const part2 = (raw) => {
  parseInput(raw);
  const paths = Array.from(
    Cave.get("start").pathsToEnd({ oneSmallRevisitAllowed: true }),
  ).filter((path) => path.endsWith(" --> end"));
  return paths.length;
};
